client/db/src/lib.rs

```diff
@@ -1320,38 +1319,67 @@ impl<Block: BlockT> Backend<Block> {
 		number: NumberFor<Block>,
 	) -> ClientResult<()> {
 		let number_u64 = number.saturated_into::<u64>();
-		if number_u64 > self.canonicalization_delay {
-			let new_canonical = number_u64 - self.canonicalization_delay;
+		if number_u64 <= self.canonicalization_delay {
+			return Ok(())
+		}
 
-			if new_canonical <= self.storage.state_db.best_canonical().unwrap_or(0) {
-				return Ok(())
-			}
-			let hash = if new_canonical == number_u64 {
+		let new_canonical = number_u64 - self.canonicalization_delay;
+		let best_canonical = self.storage.state_db.best_canonical().unwrap_or(0);
+
+		if new_canonical <= best_canonical {
+			return Ok(())
+		}
+
+		let best_number = self.blockchain.info().best_number.saturated_into();
+		// If the `best_number` is one off from the current block we are just importing, we can
+		// take the block number of the current block as the best block. Even if this block is not
+		// imported as best block, we know its hash and can canonicalize it if required below.
+		let best_number = if best_number + 1 == number_u64 { number_u64 } else { best_number };
```

Comment:
I don't get the question? Can you reformulate please?

Reply:
Would this change work instead of the +1 range:
```diff
diff --git a/client/db/src/lib.rs b/client/db/src/lib.rs
index 426876f5c..0c680d935 100644
--- a/client/db/src/lib.rs
+++ b/client/db/src/lib.rs
@@ -1547,6 +1547,8 @@ impl<Block: BlockT> Backend<Block> {
                        // VERY IMPORTANT
                        drop(operation.old_state);
 
+                       // canonicalize blocks which are old enough, regardless of finality.
+                       self.force_delayed_canonicalize(&mut transaction, hash, *header.number())?
                        if finalized {
                                // TODO: ensure best chain contains this block.
                                self.ensure_sequential_finalization(header, Some(last_finalized_hash))?;
@@ -1557,9 +1559,6 @@ impl<Block: BlockT> Backend<Block> {
                                        &mut finalization_displaced_leaves,
                                        operation.commit_state,
                                )?;
-                       } else {
-                               // canonicalize blocks which are old enough, regardless of finality.
-                               self.force_delayed_canonicalize(&mut transaction, hash, *header.number())?
                        }
 
                        if !existing_header {
```

Reply:
Ok I did not wake up properly, `ensure_sequential_finalization` is not doing the actual finalizing.
So finalization was skip in the calling process and we run it with the +1, is that the idea ?

Reply:
Ok I read the comment, make sense to me now.

Reply:
I added these changes because we have a least one test that is running with canonicalization delay of `0`. This means we canonicalize as we import a block. When we are importing X as new best block, the view of the backend is that `X - 1` is still the best, because our changes are here not yet committed. So, we are checking if the old best block is `X - 1` and if yes, we take the current block as best block. Down below where we do the force canonicalization we have this check `if to_canonicalize == number_u64` which then gives us the correct hash.